### PR TITLE
Second attempt at OkHttp Zipkin integration

### PR DESCRIPTION
**Pull request by yschimke**
Now that there is a Event Listener, I took a stab at zipkin integration again.  I didn't find it useful to use most of the existing integration with either Call.Factory or dispatcher etc, but advice welcome.  Each listener instance seems like a useful holder for the thread local context etc.

```
$ cat ~/.zipkinrc
SENDER=http://localhost:9411/api/v1/spans
DISPLAY=http://localhost:9411/zipkin/traces/{traceid}

$ ./oksocial --zipkinTrace 'https://httpbin.org/redirect-to?url=http://twitter.com/'
```
image

cc @adriancole - advice welcome, but no problem if you are busy.

### CODE REVIEW

**Comment by yschimke** [comment]
@adriancole - if you look into this I have another couple of questions

1. Is there a convention for finding the configured zipkin sink? Either a URL or whether scribe is configured correctly? ENV variables? url patterns (scribe:/file, http://zipkin:9411/api)?
2. Is there any helper code for generating a link to open the resulting zipkin trace in a browser?

File: pom.xml [hunk 1]
```diff
@@ -68,6 +68,36 @@
...
+      <version>1.0.0</version>
+    </dependency>
+    <dependency>
+      <groupId>io.zipkin.brave</groupId>
+      <artifactId>brave-okhttp</artifactId>
+      <version>4.5.1</version>
+    </dependency>
+    <dependency>
+      <groupId>io.zipkin.brave</groupId>
+      <artifactId>brave-spancollector-http</artifactId>
```
Comment (adriancole) [inline pom.xml]: think you don't need this

File: pom.xml [hunk 2]
```diff
@@ -68,6 +68,36 @@
...
+      <version>4.5.1</version>
+    </dependency>
+    <dependency>
+      <groupId>io.zipkin.brave</groupId>
+      <artifactId>brave-instrumentation-okhttp3</artifactId>
+      <version>4.5.1</version>
+    </dependency>
+    <dependency>
+      <groupId>io.zipkin.brave</groupId>
+      <artifactId>brave-okhttp</artifactId>
```
Comment (adriancole) [inline pom.xml]: also don't need this

File: src/main/java/com/baulsupp/oksocial/Main.java [hunk 3]
```diff
@@ -44,10 +49,13 @@
 import com.baulsupp.oksocial.security.OpenSCUtil;
 import com.baulsupp.oksocial.services.twitter.TwitterCachingInterceptor;
 import com.baulsupp.oksocial.services.twitter.TwitterDeflatedResponseInterceptor;
+import com.baulsupp.oksocial.tracing.ZipkinTracingInterceptor;
+import com.baulsupp.oksocial.tracing.ZipkinTracingListener;
 import com.baulsupp.oksocial.util.FileContent;
 import com.baulsupp.oksocial.util.InetAddressParam;
 import com.baulsupp.oksocial.util.LoggingUtil;
 import com.baulsupp.oksocial.util.ProtocolUtil;
+import com.github.kristofa.brave.LoggingReporter;
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/Main.java]: scrub anything that starts w/com.github.kristofa

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  @Override public Response intercept(Chain chain) throws IOException {
+    Request request = chain.request();
+
+    TraceContext traceContext = tracing.currentTraceContext().get();
+    if (traceContext != null) {
+      Request.Builder newRequest = request.newBuilder();
+
+      TraceContext.Injector<Request.Builder> i =
+          B3Propagation.create(Propagation.KeyFactory.STRING).injector(
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingInterceptor.java]: you can use tracing.propagation() and save the injector off as a field. ack that in this case we are likely looking at a request-specific interceptor

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  public ZipkinTracingInterceptor(Tracing tracing) {
+
+    this.tracing = tracing;
+  }
+
+  @Override public Response intercept(Chain chain) throws IOException {
+    Request request = chain.request();
+
+    TraceContext traceContext = tracing.currentTraceContext().get();
+    if (traceContext != null) {
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingInterceptor.java]: obv style, but if doing like this, I'd early return and comment we are only resuming traces not starting new ones

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    callSpan.tag(TraceKeys.HTTP_PATH, call.request().url().encodedPath());
+    callSpan.tag(TraceKeys.HTTP_METHOD, call.request().method());
+    callSpan.tag(TraceKeys.HTTP_HOST, call.request().url().host());
+    callSpan.kind(Span.Kind.CLIENT);
+
+    spanInScope = tracer.withSpanInScope(callSpan);
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: probably want to set this in scope even in noop bc you get consistent ids in logs if using SLF4J

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  @Override public void fetchEnd(Call call, Throwable throwable) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    callSpan.finish();
+    spanInScope.close();
+
+    if (throwable != null) {
+      callSpan.annotate(throwable.toString());
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: usually we add the tag "error" vs annotate (if it is unrecoverable), and something with the word "error" in it as an annotation if it is recoverable.

tag("error" colors the span red
annotate("...error..." colors the span yellow

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    }
+  }
+
+  @Override public void dnsStart(Call call, String domainName) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    dnsSpan =
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: you could consider stacking the span here, ex putting the dns one in scope (again for log correlation)

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+  @Override
+  public void dnsEnd(Call call, String domainName, @Nullable List<InetAddress> inetAddressList,
+      @Nullable Throwable throwable) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    if (throwable == null) {
+      dnsSpan.tag("dns results",
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: space is annoying to search with :) maybe dns.results

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+          inetAddressList.stream().map(Object::toString).collect(Collectors.joining(", ")));
+    } else {
+      dnsSpan.annotate(throwable.toString());
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: same comment about leveraging the word "error" https://github.com/openzipkin/zipkin-api/blob/master/thrift/zipkinCore.thrift#L218

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    connectionEvent = true;
+  }
+
+  @Override public void connectionFound(Call call, Connection connection) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    if (!connectionEvent) {
+      tracer.newChild(callSpan.context())
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: could consider annotate here

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+    requestSpan.finish();
+  }
+
+  @Override public void responseHeadersStart(Call call) {
+    if (callSpan.isNoop()) {
+      return;
+    }
+
+    responseSpan =
```
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: so this is a span for reading the response. It probably makes sense to close the call span at this point as usually RPC span ends when headers are read (definitely no single rule here)

File: src/main/java/com/baulsupp/oksocial/Main.java (hunk 3 above)
Comment (yschimke) [inline src/main/java/com/baulsupp/oksocial/Main.java]: Does LoggingReporter exist anymore?

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java (hunk 8 above)
Comment (yschimke) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: what would the change here be?

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java (hunk 12 above)
Comment (yschimke) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: Would I need another parent span above call to see them together?

File: src/main/java/com/baulsupp/oksocial/Main.java [hunk 13]
```diff
@@ -501,6 +514,38 @@ public void initialise() throws Exception {
...
+          .sampler(Sampler.ALWAYS_SAMPLE)
+          .build();
+
+      HttpTracing httpTracing = HttpTracing.create(tracing);
+
+      Tracer tracer = tracing.tracer();
+
+      Consumer<TraceContext> opener =
+          tc -> closeables.add(
+              () -> openLink("http://localhost:9411/zipkin/traces/" + tc.traceIdString()));
```
Comment (yschimke) [inline src/main/java/com/baulsupp/oksocial/Main.java]: any advice on this?

Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/Main.java]: if the goal is to see the trace, you could hit /api/v1/trace/+id, maybe using something to pretty-print it

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java (hunk 8 above)
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: so you could do two things: you could make the spanInScope a stack of some sort and push a result of tracer.withSpanInScope onto it here. Another would be more direct which is to save another ref like dnsSpanInScope, which is closed on dnsEnd

File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java (hunk 12 above)
Comment (adriancole) [inline src/main/java/com/baulsupp/oksocial/tracing/ZipkinTracingListener.java]: well they are in the same trace, so you'd see them together, just the calling span would have a critical path shorter than the response span. This is a judgement call.

[][][][][][][][] <- request span
              [][][][][] <- response reading span

File: src/main/java/com/baulsupp/oksocial/Main.java (hunk 13 above)
Comment (yschimke) [inline src/main/java/com/baulsupp/oksocial/Main.java]: No, happy to spawn a browser.  But my question was really more around 

https://github.com/yschimke/oksocial/pull/255#issuecomment-315769686

I'm wondering about something like

$ cat ~/.zipkinrc
ZIPKIN_SINK=http://localhost:8411/api/v1/trace
ZIPKIN_UI=http://localhost:8411/zipkin/traces/{traceid}

vs

$ cat ~/.zipkinrc
ZIPKIN_SINK=scribe://localhost:1463
ZIPKIN_UI=https://zipkin.company.com/zipkin/traces/{traceid}

**Comment by adriancole** [comment]
> Is there a convention for finding the configured zipkin sink? Either a URL or whether scribe is configured correctly? ENV variables? url patterns (scribe:/file, http://zipkin:9411/api)?
often there's nested properties or similar like so:
https://github.com/openzipkin/zipkin-finagle

> Is there any helper code for generating a link to open the resulting zipkin trace in a browser?
I've not seen any, especially as you won't always know what the base URL is (can be different as some sites don't host the UI on http collectors)

**Comment by adriancole** [comment]
gotcha https://github.com/yschimke/oksocial/pull/255#issuecomment-317327118

**Comment by yschimke** [comment]
Awesome, thanks.

I'll put those in a known rc file to avoid retyping on each command.  Read default impls based on those system properties or rc file config.

Generally seems like having a URI for the Zipkin sink would be a useful feature for Brave. e.g. https://github.com/rsocket/rsocket-java/blob/1.0.x/rsocket-core/src/main/java/io/rsocket/uri/UriTransportRegistry.java

**Comment by yschimke** [comment]
$ ./oksocial --zipkinTrace http://httpbin.org/redirect-to?url=http%3A%2F%2Fgoogle.com%2Frobots.txt

image

$ ./oksocial -z http://httpbin.org/redirect-to?url=http%3A%2F%2Fgoogle.com%2Frobots.txt

image

**Comment by adriancole** [comment]
Very cool. I will play with this tomorrow cc @anuraaga @devinsba (just
randomly picking people who might find this neat)

**Comment by yschimke** [comment]
@adriancole n.b. It needs this PR of okhttp https://github.com/square/okhttp/pull/3476

There is some terrible code for loading config from ~/.zipkinrc.  Feel free to steal if useful.

**Comment by yschimke** [comment]
@adriancole this is now released on top of OkHttp 3.9.0.  Be interested in any feedback from test use.

```
$ brew install yschimke/tap/oksocial
$ cat ~/.zipkinrc
SENDER=http://localhost:9411/api/v1/spans
DISPLAY=http://localhost:9411/zipkin/traces/{traceid}
$ ./oksocial --authorize twitter
$ ./oksocial --zipkinTrace 'https://httpbin.org/redirect-to?url=http://twitter.com/' --dns netty
```

**Comment by adriancole** [comment]
congratulations!

**Pull request merged**